xbmc/interfaces/json-rpc/PlayerOperations.cpp

```diff
@@ -478,6 +478,45 @@ JSONRPC_STATUS CPlayerOperations::SetSpeed(const std::string &method, ITransport
   }
 }
 
+JSONRPC_STATUS CPlayerOperations::SetTempo(const std::string& method,
+                                           ITransportLayer* transport,
+                                           IClient* client,
+                                           const CVariant& parameterObject,
+                                           CVariant& result)
+{
+  switch (GetPlayer(parameterObject["playerid"]))
+  {
+    case Video:
+    case Audio:
+    {
+      auto& components = CServiceBroker::GetAppComponents();
+      const auto appPlayer = components.GetComponent<CApplicationPlayer>();
+      if (!appPlayer->SupportsTempo() || appPlayer->IsPausedPlayback())
+        return FailedToExecute;
+
+      if (parameterObject["tempo"].isDouble())
+        appPlayer->SetTempo(parameterObject["tempo"].asFloat());
+      else if (parameterObject["tempo"].isString())
+      {
+        if (parameterObject["tempo"].asString().compare("increment") == 0)
+          CBuiltins::GetInstance().Execute("playercontrol(tempoup)");
```

Comment:
CBuiltins is a GUI class (used for actions from input and skins). Isnt it possible to use the app player directly?
Yeah I know other methods already use this, doesn't necessarily mean it's right

Reply:
that would mean copy-pasting that logic. Or what would be a good place to extract it to?

Reply:
Isn't there equivalent methods on the interface? If not I guess the logic for up/down could be centralised in playerutils so that the raw methods with the actual tempo values end up in the interface calls

Reply:
> Isn't there equivalent methods on the interface?

neither in CPlayerUtils nor in CApplicationPlayer

> the logic for up/down could be centralised in playerutils so that the raw methods with the actual tempo values end up in the interface calls

thanks! will try